File: app/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

public enum MapDirection {
    NORTH,
    SOUTH,
    WEST,
    EAST;

    static private final Vector2d northUnitVector = new Vector2d(0, 1);
    static private final Vector2d southUnitVector = new Vector2d(0, -1);
    static private final Vector2d westUnitVector = new Vector2d(-1, 0);
    static private final Vector2d eastUnitVector = new Vector2d(1, 0);

    public String toString() {
        return switch (this) {
            case NORTH -> "Północ";
            case SOUTH -> "Południe";
            case WEST -> "Zachód";
            case EAST -> "Wschód";
        };
    }

    public MapDirection next() {
        return switch (this) {
            case NORTH -> EAST;
            case SOUTH -> WEST;
            case WEST -> NORTH;
            case EAST -> SOUTH;
        };
    }

    public MapDirection previous() {
        return switch (this) {
            case NORTH -> WEST;
            case SOUTH -> EAST;
            case WEST -> SOUTH;
            case EAST -> NORTH;
        };
    }

    public Vector2d toUnitVector() {
        return switch (this) {
            case NORTH -> northUnitVector;
            case SOUTH -> southUnitVector;
            case WEST -> westUnitVector;
            case EAST -> eastUnitVector;
        };   
    }
}

File: kotlin/app/src/main/kotlin/agh/ics/oop/model/MapDirection.kt
package agh.ics.oop.model

enum class MapDirection {
    NORTH,
    EAST,
    SOUTH,
    WEST;

    override fun toString(): String = when (this) {
        NORTH -> "Północ"
        EAST -> "Wschód"
        SOUTH -> "Południe"
        WEST -> "Zachód"
    }

    fun next(): MapDirection = values()[(ordinal + 1) % values().size]

    fun previous(): MapDirection = values()[(ordinal - 1 + values().size) % values().size]
}